### PR TITLE
plan: Next skeleton

### PR DESCRIPTION
**Pull request by coocood**
Just added the method, not implemented yet.

`Do` will be kept until all plans implemented `Next`

### CODE REVIEW

File: plan/plan.go [hunk 1]
```diff
@@ -40,6 +41,10 @@ type Plan interface {
 	// If index can be used, a new index plan is returned, 'filtered' is true.
 	// If no index can be used, the original plan is returned and 'filtered' return false.
 	Filter(ctx context.Context, expr expression.Expression) (p Plan, filtered bool, err error)
+
+	// Next returns the next row of data and rowKeys, nil data means there is no more data to return.
+	// Aggregation plan will fetch all the data at the first call.
+	Next(ctx context.Context) (data []interface{}, rowKeys []*RowKeyEntry, err error)
```
Comment (siddontang) [inline plan/plan.go]: I think we should define a Row struct as the return type, e.g, if we have to add another xxx, we can change in Row

```
type Row struct {
    Data []interface{}
    Keys []*RowKeyEntry
    // other 
}
```

**Comment by ngaut** [comment]
We need to discuss the PR later, please do not merge too fast.

File: plan/plan.go [hunk 2]
```diff
@@ -40,6 +41,10 @@ type Plan interface {
 	// If index can be used, a new index plan is returned, 'filtered' is true.
 	// If no index can be used, the original plan is returned and 'filtered' return false.
 	Filter(ctx context.Context, expr expression.Expression) (p Plan, filtered bool, err error)
+
+	// Next returns the next row of data and rowKeys, nil data means there is no more data to return.
+	// Aggregation plan will fetch all the data at the first call.
+	Next(ctx context.Context) (row *Row, err error)
```
Comment (siddontang) [inline plan/plan.go]: How to terminate iteration explicitly?

Comment (coocood) [inline plan/plan.go]: How about when user terminate, we check if the plan implements `io.Closer`, if it does implement it, we call `Close`?

And we can also extract `Filter` method to `Filterer` interface, so we don't need every plan to implement it.

Comment (coocood) [inline plan/plan.go]: Or add `Close` to `Plan`, as most plans need to implement `Close` method.

Comment (siddontang) [inline plan/plan.go]: I prefer adding Close to Plan.

**Comment by coocood** [comment]
@siddontang @qiuyesuifeng PTAL

File: plan/plans/union.go [hunk 3]
```diff
@@ -139,3 +141,127 @@ func (p *UnionPlan) GetFields() []*field.ResultField {
...
+		p.rows = append(p.rows, row)
+		err = t.Set(row.Data, []interface{}{true})
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	// Eval the following select statements
+	for i := range p.Distincts {
+		err = p.fetchSrc(ctx, i, t)
```
Comment (siddontang) [inline plan/plans/union.go]: I think the src in fetchSrc doesn't match the above comment following.

File: plan/plans/where.go [hunk 4]
```diff
@@ -75,3 +77,44 @@ func (r *FilterDefaultPlan) Do(ctx context.Context, f plan.RowIterFunc) (err err
...
+	for {
+		row, err = r.Plan.Next(ctx)
+		if row == nil || err != nil {
+			return nil, errors.Trace(err)
+		}
+		r.evalArgs[expressions.ExprEvalIdentFunc] = func(name string) (interface{}, error) {
+			return getIdentValue(name, r.GetFields(), row.Data, field.DefaultFieldFlag)
+		}
+		var meet bool
+		meet, err = r.meetRequirement(ctx)
```
Comment (siddontang) [inline plan/plans/where.go]: meetCondition may be better?

File: plan/plans/join.go [hunk 5]
```diff
@@ -342,3 +347,167 @@ func appendRow(prefix []interface{}, in []interface{}) []interface{} {
...
+			row.Data = append(row.Data, make([]interface{}, rightLen)...)
+		}
+		r.matchedRows = append(r.matchedRows, row)
+	}
+	return nil
+}
+
+func (r *JoinPlan) nextCrossJoin(ctx context.Context) (row *plan.Row, err error) {
+	for {
+		if r.curRow == nil {
```
Comment (siddontang) [inline plan/plans/join.go]: https://dev.mysql.com/doc/refman/5.7/en/nested-loop-joins.html

for the common join algorithm, we should first do left Next, and travel right iteratively.

**Comment by siddontang** [comment]
LGTM

File: plan/plan.go [hunk 6]
```diff
@@ -40,6 +41,14 @@ type Plan interface {
 	// If index can be used, a new index plan is returned, 'filtered' is true.
 	// If no index can be used, the original plan is returned and 'filtered' return false.
 	Filter(ctx context.Context, expr expression.Expression) (p Plan, filtered bool, err error)
+
+	// Next returns the next row of data and rowKeys, nil data means there is no more data to return.
```
Comment (qiuyesuifeng) [inline plan/plan.go]: `rowKeys` or `row datas`?

File: plan/plan.go [hunk 7]
```diff
@@ -47,3 +56,18 @@ type Planner interface {
...
 }
+
+// Row represents a record row.
+type Row struct {
+	Data    []interface{}
+	RowKeys []*RowKeyEntry
+}
+
+// RowKeyEntry is designed for Delete statement in multi-table mode,
+// we should know which table this row comes from
```
Comment (qiuyesuifeng) [inline plan/plan.go]: End of '.'

File: plan/plans/explain.go [hunk 8]
```diff
@@ -26,6 +26,7 @@ import (
 	"github.com/pingcap/tidb/plan"
 	"github.com/pingcap/tidb/stmt"
 	"github.com/pingcap/tidb/util/format"
+	"strings"
```
Comment (qiuyesuifeng) [inline plan/plans/explain.go]: It is better to update package import, we actually import like this:
``` 
origin packages
empty line
github/other packages
```

File: plan/plans/explain.go [hunk 9]
```diff
@@ -68,3 +71,28 @@ func (r *ExplainDefaultPlan) GetFields() []*field.ResultField {
...
+
+// Next implements plan.Plan Next interface.
+func (r *ExplainDefaultPlan) Next(ctx context.Context) (row *plan.Row, err error) {
+	if r.lines == nil {
+		var buf bytes.Buffer
+		w := format.IndentFormatter(&buf, "│   ")
+		r.S.Explain(ctx, w)
+		r.lines = strings.Split(string(buf.Bytes()), "\n")
+	}
+	if r.cursor >= len(r.lines)-1 {
```
Comment (qiuyesuifeng) [inline plan/plans/explain.go]: Can use `if r.cursor == len(r.lines)` here?

File: plan/plans/from.go [hunk 10]
```diff
@@ -94,12 +95,48 @@ func (r *TableNilPlan) Do(ctx context.Context, f plan.RowIterFunc) error {
...
+	}
+	if !r.iter.Valid() || !strings.HasPrefix(r.iter.Key(), r.T.KeyPrefix()) {
+		return
+	}
+	id, err := util.DecodeHandleFromRowKey(r.iter.Key())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	rk := r.T.RecordKey(id, nil)
+	row = &plan.Row{}
```
Comment (qiuyesuifeng) [inline plan/plans/from.go]: Add some comment?

File: plan/plans/groupby.go [hunk 11]
```diff
@@ -322,3 +324,112 @@ func (r *GroupByDefaultPlan) evalEmptyTable(ctx context.Context) ([]interface{},
 	}
 	return out, nil
 }
+
+// Next implements plan.Plan Next interface.
+func (r *GroupByDefaultPlan) Next(ctx context.Context) (row *plan.Row, err error) {
+	if r.rows == nil {
+		r.rows = make([]*groupRow, 0)
```
Comment (qiuyesuifeng) [inline plan/plans/groupby.go]: r.rows  := []*groupRow{}

File: plan/plans/having.go [hunk 12]
```diff
@@ -78,3 +80,41 @@ func (r *HavingPlan) Do(ctx context.Context, f plan.RowIterFunc) (err error) {
...
+			return getIdentValue(name, r.Src.GetFields(), srcRow.Data, field.CheckFieldFlag)
+		}
+		r.evalArgs[expressions.ExprEvalPositionFunc] = func(position int) (interface{}, error) {
+			// position is in [1, len(fields)], so we must decrease 1 to get correct index
+			// TODO: check position invalidation
+			return srcRow.Data[position-1], nil
+		}
+		var v bool
+		v, err = expressions.EvalBoolExpr(ctx, r.Expr, r.evalArgs)
+
```
Comment (qiuyesuifeng) [inline plan/plans/having.go]: Remove empty line?

File: plan/plans/from.go [hunk 13]
```diff
@@ -315,3 +352,64 @@ func (r *TableDefaultPlan) Do(ctx context.Context, f plan.RowIterFunc) error {
...
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return
+}
+
+// Close implements plan.Plan Close interface.
+func (r *TableDefaultPlan) Close() error {
+	if r.iter != nil {
+		r.iter.Close()
```
Comment (qiuyesuifeng) [inline plan/plans/from.go]: Move `r.iter = nil` here?

File: rset/rsets/rsets.go [hunk 14]
```diff
@@ -53,16 +65,13 @@ func (r Recordset) Fields() (fields []*field.ResultField, err error) {
 
 // FirstRow implements rset.Recordset.
 func (r Recordset) FirstRow() (row []interface{}, err error) {
-	rows, err := r.Rows(1, 0)
-	if err != nil {
-		return nil, err
+	ro, err := r.Plan.Next(r.Ctx)
+	if ro == nil || err != nil {
+		return nil, errors.Trace(err)
```
Comment (qiuyesuifeng) [inline rset/rsets/rsets.go]: Need call `r.Plan.Close()` here?

Comment (coocood) [inline rset/rsets/rsets.go]: It's not necessary to call r.Plan.Close() but doesn't hurt.

**Comment by qiuyesuifeng** [comment]
LGTM

**Pull request merged**